shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -243,287 +218,213 @@ class JSBackend {
+          Bounds(_, _) | WithExtension(_, _) | Function(_, _) | Union(_, _) | _: Arr =>
+        throw CodeGenError(s"cannot inherit from type $ty")
+    }
+    impl(ty)
   }
 
-  private def translateClassMember(
-      method: MethodDef[Left[Term, Type]]
-  )(implicit scope: Scope): JSClassMemberDecl = {
-    val name = method.nme.name
-    method.rhs.value match {
-      case Lam(params, body) =>
-        JSClassMethod(name, translateParams(params), L(translateTerm(body)))
-      case term => JSClassGetter(name, L(translateTerm(term)))
+  /**
+    * Resolve inheritance of all declared classes.
+    */
+  protected def resolveInheritance(classSymbols: Ls[ClassSymbol]): (Ls[ClassSymbol], Ls[ClassSymbol -> ClassSymbol]) = {
+    val (noBase, relations) = classSymbols.partitionMap { derivedClass =>
+      val baseClass = resolveClassBase(derivedClass.actualType)
+      derivedClass.baseClass = baseClass
+      baseClass.map(_ -> derivedClass).toRight(derivedClass)
     }
+    val inRelations = Set.from(relations.iterator.flatMap { case (a, b) => a :: b :: Nil })
+    (noBase.filterNot { inRelations.contains(_) }, relations)
+  }
+
+  protected def translateClassDeclarations(sortedClasses: Iterable[ClassSymbol]): Ls[JSClassDecl] = {
+    sortedClasses.flatMap { sym =>
+      sym.body match {
+        case S(_) => N // Don't translate if done
+        case N =>
+          val body = translateClassDeclaration(sym)(topLevelScope)
+          sym.body = S(body)
+          S(body)
+      }
+    }.toList
   }
 }
 
 class JSWebBackend extends JSBackend {
-  // This will be changed during code generation.
-  private var hasWithConstruct = false
-
-  // Name of the helper function for `with` construction.
-  private val withConstructFnName = topLevelScope allocateJavaScriptName "withConstruct"
-
   // Name of the array that contains execution results
-  val resultsName: Str = topLevelScope allocateJavaScriptName "results"
+  val resultsName: Str = topLevelScope declareRuntimeSymbol "results"
 
-  val prettyPrinterName: Str = topLevelScope allocateJavaScriptName "prettyPrint"
+  val prettyPrinterName: Str = topLevelScope declareRuntimeSymbol "prettyPrint"
 
   polyfill.use("prettyPrint", prettyPrinterName)
 
   def apply(pgrm: Pgrm): Ls[Str] = {
     val (diags, (typeDefs, otherStmts)) = pgrm.desugared
 
-    // Collect type aliases into a map so we can normalize them.
-    typeDefs foreach {
-      case TypeDef(Als, TypeName(name), tparams, body, _, _) =>
-        val tnames = tparams map { case TypeName(nme) => nme }
-        typeAliasMap(name) = tnames -> body
-      case TypeDef(Trt, TypeName(name), _, _, _, _) =>
-        traitNames += name
-      case _ => ()
+    val classSymbols = declareTypeDefs(typeDefs)
+    val (isolatedClasses, relations) = resolveInheritance(classSymbols)
+    val sortedClasses = try topologicallySort(relations) catch {
+      case e: RuntimeException => throw CodeGenError("cyclic inheritance involving")
```

Comment:
> `case e: RuntimeException `

This is disgusting. It will swallow _any_ error and make it pass for a cyclic inheritance, even if it's another source or error. The better thing to do would be to throw a proper catchable exception in the topological sort function.